src/Common/MemoryBufferWriter.cs

```diff
@@ -170,5 +213,65 @@ public byte[] ToArray()
 
             return result;
         }
+
+        public override void Flush() { }
+        public override Task FlushAsync(CancellationToken cancellationToken) => Task.CompletedTask;
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        public override void WriteByte(byte value)
+        {
+            if (_currentSegment != null && _position < _currentSegment.Length)
```

Comment:
Skip bounds checks
```csharp
var currentSegment = _currentSegment;
var position = _position;
if (currentSegment != null && (uint)position < (uint)currentSegment.Length)
{
    currentSegment[position] = value;
    _position = position + 1;
}
```

Reply:
This actually seemed to make things slower... Trying again.

Reply:
You need both the array and index in a local rather than working off a memory location (`this._currentSegment`) for it to work; otherwise the Jit won't assume the values are the same when you use them to do the indexing.

Reply:
It was better for arrays that were memory references at one stage; but then it was made worse 😢 https://github.com/dotnet/coreclr/pull/15756

Reply:
Yea, it's worse